src/linear_algebra/affine_space.lean

```diff
@@ -523,6 +510,21 @@ that subspace's direction yields the original subspace. -/
 ext_of_direction_eq (direction_mk' p s.direction)
                     ⟨p, set.mem_inter (self_mem_mk' _ _) hp⟩
 
+/-- If an affine subspace contains a set of points, it contains the
+`span_points` of that set. -/
+lemma span_points_subset_coe_of_subset_coe {s : set P} {s1 : affine_subspace k V P} (h : s ⊆ s1) :
```

Comment:
I think that `affine_span` or `span_points` forms a galois connection (or galois insertion?) with `coe`. There is a very good API for such connections, and it allows you to quickly prove a lot of things for "free".

Reply:
Yes, that's on my (long) list of basic things about affine subspaces that still need to be defined / proved, having been mentioned in #2816. It appears `submodule.gi` is used to prove four lemmas (`span_empty`, `span_univ`, `span_union`, `span_Union`), all of which have analogues for affine subspaces that can be proved in the same way. (While `span_points_subset_coe_of_subset_coe` is one of the pieces that would be used in constructing the `galois_insertion`.)

Reply:
Aha, I thought that it could also be used for getting lattice instances. Can
```
src/order/galois_connection.lean:def lift_complete_lattice
```
make your life easier? It might save number of lines for the complete lattice instance below.

Reply:
It looks like it would be a pain to arrange the `choice` function to get definitional equality for `inf_coe`; at least, using a trivial `choice` definition as in `submodule.gi` would result in an extra level of `affine_span` involved in the definitions of `inf` and `Inf`. Defining the whole `complete_lattice` structure at once felt simpler than approaches involving defining just part of the structure (e.g. the `partial_order`), then defining / proving more things using those order definitions, then defining the rest of the `complete_lattice` structure.

Reply:
@jcommelin For `sub*` structures I prefer to define `complete_lattice` as it done for `submonoid`s: define `has_Inf`, `\bot`, `\top`, `inf`, then get the rest for free from `complete_lattice_of_Inf`. In particular, I'd define `affine_span` in the same way as `submonoid.closure`.

Reply:
The logic behind the present definition of `affine_span` is to treat affine spaces as a variation on modules as far as possible and define things in terms of the corresponding definitions for modules as far as that makes sense, so hopefully reducing the amount of reproving needed whenever results for modules become directly applicable; thus, to define `affine_span` in terms of `submodule.span`. (The reason `affine_independent` is defined directly then proved equivalent to something in terms of `linear_independent` is essentially that the subtypes involved in using `linear_independent` there are a pain to deal with.) Defining `affine_span` using `Inf` (as opposed to treating equality to an `Inf` as a lemma to be proved, which this PR provides most of the pieces needed to prove) would probably have the effect of shifting complexity around, from defining the lattice structure to proving how definitions such as `affine_span` relate to the corresponding definitions for submodules.

Reply:
Ok, I'm convinced!